USB: Automatically define COUNT number of USB Devices

finikorg opened the pull request:
Use UTIL_LISTIFY to automatically define COUNT number of instances of HID and CDC ACM.

finikorg committed:
usb: Add Kconfig template for DEVICE_COUNT

DEVICE_COUNT for some USB classes means how many instances of the
class can be supported by USB stack.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: hid: Use UTIL_LISTIFY and device count macro

Use UTIL_LISTIFY and Kconfig device count template for specify number
of HID Devices.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: usb: Correct USB HID device name

Use "HID_0" as a device name since we have only device count now.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: cdc_acm: Use UTIL_LISTIFY and device count macro

Use new method for defining USB CDC ACM devices using UTIL_LISTIFY and
device count from menuconfig.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: cdc_acm: Update CDC ACM Device name

Update device name for USB CDC ACM.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -1074,30 +1074,33 @@ static const struct uart_driver_api cdc_acm_driver_api = {
+			    &cdc_acm_driver_api);
+
+/* Use Automatic endpoint assignment */
+#define AUTO_EP_IN	0x80
```
@jfischer-phytec-iot OK, what about assigning those default values to AUTO_EP_IN ?

jfischer-phytec-iot replied:
Up to you, but do not assign the same endpoint address for more than one endpoint, you can rename the macros above or just use it:

```c
#define DEFINE_CDC_ACM_DESCR_AUTO(x, _) \
	DEFINE_CDC_ACM_DESCR(x, CDC_ACM_INT_EP_ADDR, CDC_ACM_OUT_EP_ADDR, CDC_ACM_IN_EP_ADDR);

#define DEFINE_CDC_ACM_EP_AUTO(x, _) \
	DEFINE_CDC_ACM_EP(x, CDC_ACM_INT_EP_ADDR, CDC_ACM_OUT_EP_ADDR, CDC_ACM_IN_EP_ADDR);
```

jfischer-phytec-iot replied:
@finikorg You do not need AUTO_EP_IN and AUTO_EP_OUT macros, the default addresses are already [defined,](https://github.com/zephyrproject-rtos/zephyr/pull/13310/commits/a23cb10f29f4a499c88fe76c093b73b11a2fa0a5#diff-e33cccf141c95f57405ce187a047a007R76) see my comment above.

finikorg replied:
@jfischer-phytec-iot I forgot to delete these definition. Do you want to keep those definitions?

jfischer-phytec-iot replied:
@finikorg I would like to keep them, but you can change default values if you want.

finikorg replied:
@jfischer-phytec-iot I think we should use auto endpoint configuration especially those default values are not correct for all platform ( for example https://github.com/zephyrproject-rtos/zephyr/blob/master/drivers/usb/device/usb_dc_kinetis.c#L284)

jfischer-phytec-iot replied:
@finikorg The default value of a endpoint address  will be reassigned, it just have to be valid endpoint address and unique inside the interface. In the quoted kinetis driver, the half of the endpoints is just pre-selected as IN.

finikorg replied:
@jfischer-phytec-iot I have test where you can see that proposed way works just fine, could you check this last commit 46b6e56b193f376f1d5f6f08c0a365a57af272e8.
You can run it in native_posix and it will print endpoint assigned.

finikorg committed:
tests: usb: Add endpoint allocation verification test

Add verification that interface number and endpoint addresses are
correctly allocated.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
samples: hid-cdc: Update USB device names and project conf

Update project configuration and USB HID and CDC ACM device names for
sample hid-cdc.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: device: Code cleanup

Add missing braces and simple cleanup.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: Remove usb_dc_ep_set_callback return code

Make usb_dc_ep_set_callback() return void since the code is never
used.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: dc: native_posix: Increase number of endpoints supported

Support 8 endpoints in each direction for native_posix.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg committed:
usb: tests: Add second IN endpoint to test

Demonstrate automatic endpoint assignment in tests for similar
endpoint types.

Signed-off-by: Andrei Emeltchenko <andrei.emeltchenko@intel.com>

finikorg commented:
I need to verify that everything works with this PR and other PRs already applied. (added DNM for now)

finikorg commented:
Rebased against recent master

finikorg commented:
Rebased

finikorg commented:
Rebased and fixed conflicts

finikorg commented:
@jfischer-phytec-iot Could you check this PR again?

jfischer-phytec-iot commented on subsys/usb/class/cdc_acm.c:
```diff
@@ -67,23 +67,16 @@ LOG_MODULE_REGISTER(usb_cdc_acm);
 /* Serial state notification timeout */
 #define CDC_CONTROL_SERIAL_STATE_TIMEOUT_US 100000
 
-#define CDC_ACM_INT_EP_ADDR		0x85
```
Can you please merge it to https://github.com/zephyrproject-rtos/zephyr/pull/13310/commits/3bf084095de588c6caf2915d23eb58553dc9d55d and use the same schema for AUTO_EP_IN, AUTO_EP_INT, AUTO_EP_OUT as in  `subsys/usb/class/hid/core.c` below.
